Highway Delite TRV10 & RSF logs

highwaydelite opened the pull request:
Highway Delite TRV10 logs and RSF logs submission.

JyotiSharma0711 commented:
# TRV10/HighwayDelite/RSF - Issues and Constraints

## Flow 1

### Settle
- `context/location/city` must be an **object**.
---

## Flow 2

### Recon
- `Message/orders/0/settlements/0/id` - **object field is missing**.
- bap_uri mismatch between recon and on_recon
- `Message/orders/0/id` - **object field is missing**.
- `Message/orders/0/settlements/0/status` must be **one of the allowed values** (`PENDING`, `SETTLED`, `TO-BE-INITIATED`).
- `Settlement updated_at` **timestamp cannot be greater than** `context timestamp`.
- `Message/orders/0/settlements/0/settlement_ref_no` - is an **additional field**, as the status is `"NOT-SETTLED"`. 
- **All values should be in "0.00" format**.

### On_recon
- `Context/city_code` is **mismatched** in both the `recon` and `on_recon` call.
- `message/orders/0/settlements/0/status` must be **one of the allowed values** (`PENDING`, `SETTLED`, `TO-BE-INITIATED`).

---

## Flow 3

### Recon
- `Message/orders/0/settlements/0/id` is **mismatched** in both the `recon` and `on_recon` call.
- `Message/orders/0/settlements/0/status` must be **one of the allowed values** (`PENDING`, `SETTLED`, `TO-BE-INITIATED`).
- `Message/orders/0/settlements/0/settlement_ref_no` - is an **additional field**, as the status is `"NOT-SETTLED"`. 
- **All values should be in "0.00" format**.

### On_recon
- `Context/city_code` is **mismatched** in both the `recon` and `on_recon` call.
- `Context/timestamp` of the `on_recon` call **should be greater than** the `context/timestamp` of the `recon` call.
---

## Flow 4

### Settle and On_settle
- `Message/collecter_app_id` is an **additional field**.
- `Message/receiver_app_id` is an **additional field**.
- `Message/settlement/orders/0/inter_participant` is an **additional field**.
- `Message/settlement/orders/0/id` is an **additional field**.
- `Message/settlement/orders/0/collector` is an **additional field**.

---

## Flow 5

### Settle and On_settle
- `Message/collecter_app_id` is an **additional field**.
- `Message/receiver_app_id` is an **additional field**.

highwaydelite commented:
@JyotiSharma0711 RSF logs have been corrected.
Are the TRV10 logs verified. Please provide appropriate comment regards to TRV10.

JyotiSharma0711 commented:
# TRV10/HighwayDelite/RSF - Issues and Constraints

## Flow 1

### Settle AND On_settle
- `Message/settlement/orders/0/id` is not in valid format.

### On_report
- `Message/settlement/orders/0/id` is not in valid format.
---

## Flow 2

### Recon
- `Message/orders/0/id` is not in valid format.
- `Message/orders/0/settlements/0/status` must be **one of the allowed values** (`PENDING`, `SETTLED`, `TO-BE-INITIATED`).

### ON_Recon
- `Message/orders/0/id` is not in valid format.

---

JyotiSharma0711 commented:
@highwaydelite 

Logs are approved for RSF 2.0.0

highwaydelite commented:
Thank you @JyotiSharma0711 .
Can you verify the logs for TRV10 buyer flows.

sourabhsaini11 commented:
@highwaydelite
init
- if collector is BAP then params shouldn't be present at payments
- tag SETTLEMENT_BASIS value can't be UPI

confirm
- fulfillments.type is sent additionally

highwaydelite commented:
@sourabhsaini11 We have made the following changes
-changed the collector from BAP to BPP in init
-changed SETTLEMENT_BASIS to DELIVERY
-removed fulfillments.type in confirm

highwaydelite commented:
@sourabhsaini11 Please review and approve the updated logs.

sourabhsaini11 commented:
@highwaydelite
search
- tags SETTLEMENT_BASIS, MANDATORY_ARBITRATION & COURT_JURISDICTION should be sent first by an on_action call

confirm
- multiple SETTLEMENT_AMOUNT tags present
- payments.params is incomplete, missing bank_account_number, bank_code

on_search(Flow-1)
- message_id should match search's message_id
- invalid JSON, found payload for on_confirm

select(Flow-1)
- timestamp can't be less than past call's timestamp

Note:
- Provide only context & message as part of payload

highwaydelite commented:
@92shreyansh Please add IGM tag to this PR.

highwaydelite commented:
@sourabhsaini11 Logs have been updated for TRV10 and IGM logs have been added for verification.

amitsinha07 commented:
@highwaydelite 

# Flow 1

## Issue_1
- `Message/issue/actors/1/type` must be equal to one of the allowed values:  
  `(CONSUMER, INTERFACING_NP, COUNTERPARTY_NP, PROVIDER, AGENT, INTERFACING_NP_GRO, COUNTERPARTY_NP_GRO, CASCADED_NP_GRO, CASCADED_NP).`
- `Source_id`: `source_id (Source1)` should match the id of the `CONSUMER` actor `(CUST1)`.

## On_issue_1
- `Message` must have required property `'updated_target'`.
- `Message/issue/actors/1/type` must be one of the allowed values.
- `Message/issue/actors/2/type` must be one of the allowed values.
- `TTL = PT30S` as per the API Contract.
- `Timestamp_validation`:
  - `Issue updated_at` should be equal to or greater than `context timestamp`.
  - `Message/issue/actions/1/updated_at` should be less than or equal to `message/issue/updated_at`.

## On_issue_2
- `Context/timestamp` of `on_issue_2` should not match with `on_issue_1`.
- `Message` must have required property `'updated_target'`.
- `Message/issue/actors/1/type` and `Message/issue/actors/2/type` must be valid.
- `TTL = PT30S` as per the API Contract.
- `Timestamp_validation`: `Issue updated_at` should be equal to or greater than `context timestamp`.

---

## Flow 2

## Issue_1
- Must have required property `'message'`.
- `Context` must have required properties:
  - `'domain'`
  - `'location'`
  - `'core_version'`
  - `'bap_id'`
  - `'bap_uri'`
  - `'bpp_id'`
  - `'bpp_uri'`
  - `'timestamp'`
  - `'ttl'`
- `Message/issue/created_at` - object field is missing.

## On_issue_1
- `Message` must have required property `'updated_target'`.
- `Message/issue/actors/1/type` and `Message/issue/actors/2/type` must be valid.
- `TTL = PT30S` as per the API Contract.
- `Message/issue/actions/1/updated_at` should be less than or equal to `message/issue/updated_at`.

## On_issue_2
- `Context/timestamp` of `on_issue_2` should not match with `on_issue_1`.
- `Message` must have required property `'updated_target'`.
- `Message/issue/actors/1/type` and `Message/issue/actors/2/type` must be valid.
- `TTL = PT30S` as per the API Contract.
- `Message/issue/actions/2/updated_at` should be less than or equal to `message/issue/updated_at`.

## Issue_2
- `Context/timestamp` of `issue_2` should not match with `on_issue_2`.
- `Message/issue` must have required properties:
  - `'actors'`
  - `'level'` must be one of `(ISSUE, GRIEVANCE, DISPUTE)`.
  - `'status'` should be `Processing`.
- `Actors` must be an array.
- Additional fields:
  - `Message/issue/actions/0/description/name`
  - `Message/issue/actions/0/description/images`
  - `Message/issue/actions/0/ref_id`
  - `Message/issue/actions/0/ref_type`
  - ... (similar validations for multiple actions)

## On_issue_3
- `Message` must have required property `'updated_target'`.
- `Message/issue` must have required property `'actors'`.
- `Message/issue/level` must be one of `(ISSUE, GRIEVANCE, DISPUTE)`.
- `Message/issue/actions/4/description` must have required property `'short_desc'`.
- `TTL = PT30S` as per the API Contract.
- `Timestamp_validation`: `Issue updated_at` should be equal to or greater than `context timestamp`.
- `Action_4_desc_short`: `description short_desc` is required.
- `Message/issue/actions/4/updated_at` should be less than or equal to `message/issue/updated_at`.

## Issue_4
- `Context/timestamp` of `issue_4` should not match with `on_issue_3`.
- `Message/issue` must have required properties:
  - `'actors'`
  - `'level'` must be one of `(ISSUE, GRIEVANCE, DISPUTE)`.
  - `Escalation`: By `issue_4`, if status is still `OPEN`, `escalation_level` might be expected.
- `Actors` must be an array.
- `Action_5_desc_short`: `description short_desc` is required.
- Additional fields:
  - `Message/issue/actions/0/description/name`
  - `Message/issue/actions/0/description/images`
  - `Message/issue/actions/0/ref_id`
  - `Message/issue/actions/0/ref_type`
  - ... (similar validations for multiple actions)

## On_issue_4
- `Context/timestamp` of `on_issue_4` should not match with `issue_4`.
- `Message` must have required property `'updated_target'`.
- `Message/issue` must have required properties:
  - `'actors'`
  - `'level'` must be one of `(ISSUE, GRIEVANCE, DISPUTE)`.
  - `Message/issue/actions/2/description` must have required property `'short_desc'`.
- `TTL = PT30S` as per the API Contract.
- `Timestamp_validation`: `Issue updated_at` should be equal to or greater than `context timestamp`.
- `Message_id`: `Message ID should be unique, but 85655939-7a87-4077-92bf-c09a40d5bc48 was already used`.
- `Action_2_desc_short`: `description short_desc` is required.
- `Message/issue/actions/6/updated_at` should be less than or equal to `message/issue/updated_at`.

## Issue_4
- `Context/message_id` of the `Issue_4` call should not match with `on_issue_4`.
- `Message/issue` must have required properties:
  - `'actors'`
  - `'level'` must be one of `(ISSUE, GRIEVANCE, DISPUTE)`.
- `Action_5_desc_short`: `description short_desc` is required.
- `Actors` must be an array.
- Additional fields:
  - `Message/issue/actions/0/description/name`
  - `Message/issue/actions/0/description/images`
  - `Message/issue/actions/0/ref_id`
  - `Message/issue/actions/0/ref_type`
  - ... (similar validations for multiple actions)

highwaydelite commented:
@amitsinha07 Updated the IGM logs. Please verify at your earliest convenience.

amitsinha07 commented:
@adarshsingh87 
Regarding logs validation errors, you can skip these errors as they are false errors. I will fix those from my end.  

For mock server doubts, kindly join tomorrow's 11 AM Community Meeting. The link is given below:  

### ONDC Tech Community Call (API Implementation - All domains)  
📅 **Thursday, 20 March**  
⏰ **11:00 AM – 12:00 PM** (Time zone: Asia/Kolkata)  

🔗 **Google Meet joining info:**  
[Video call link](https://meet.google.com/bpo-pwwh-fja)  

📞 Or dial: **(US) +1 347-389-1685** PIN: **650 319 046**#  
📍 More phone numbers: [Click here](https://tel.meet/bpo-pwwh-fja?pin=6443347012985)

amitsinha07 commented:
@highwaydelite 

# FLOW 1  

## issue(2)  
- **schemaErr1**: `/message/issue/actors/2/type must be equal to one of the allowed values ( COUNTERPARTY_NP)`  
- **Updated At Match**: `Message/issue/actions/0/updated_at should match with the message/issue/created_at.`  
- **Additional Fields**:  
  - `Message/issue/actions/0/description/images is an additional field.`  
  - `Message/issue/actions/1/description/images is an additional field.`  
  - `Message/issue/actions/2/description/images is an additional field.`  
- **Updated At Order**: `Message/issue/actions/3/updated_at should be less than or equal to message/issue/updated_at.`  

## issue(3)  
- **schemaErr0**: `/message/issue/actors/2/type must be equal to one of the allowed values (COUNTERPARTY_NP)`  
- **Updated At Match**: `Message/issue/actions/0/updated_at should match with the message/issue/created_at.`  
- **Updated At Order**: `Message/issue/actions/6/updated_at should be less than or equal to message/issue/updated_at.`  
- **Org Name Mismatch**: `Message/issue/actors/2/info/org/name is mismatched with the bpp_id.`  
- **Description Code**: `Message/issue/actions/1/description/code should be “Processing”.`  

## issue_closed(4)  
- **schemaErr0**: `/message/issue/actors/2/type must be equal to one of the allowed values (COUNTERPARTY_NP)`  
- **Updated At Match**: `Message/issue/actions/0/updated_at should match with the message/issue/created_at.`  
- **Updated At Order**: `Message/issue/actions/8/updated_at should be less than or equal to message/issue/updated_at.`  
- **Org Name Mismatch**: `Message/issue/actors/2/info/org/name is mismatched with the bpp_id.`  
- **Description Code**: `Message/issue/actions/1/description/code should be “Processing”.`  


# FLOW 2  

## issue(0)  
- **schemaErr0**: `/message/issue/description/additional_desc/url must match format "uri"`  
- **Missing References**:  
  - `Reference type 'TRANSACTION' is mandatory for FLOW_2 but is missing.`  
  - `Reference type 'PROVIDER' is mandatory for FLOW_2 but is missing.`  
- **Actor Org Name Mismatch**: `For actor type "INTERFACING_NP" in FLOW_2, org.name should be "igm-mock-service.ondc.org/igm-protocol-server-preprod/ondc::ONDC:RET10"`  

## issue(2)  
- **timestamp_validation**: `Issue updated_at should be equal to or greater than context timestamp.`  
- **status_progress**: `With 4 actions recorded, status might be expected to progress beyond OPEN.`  
- **Actor Org Name Mismatch**: `For actor type "INTERFACING_NP" in FLOW_2, org.name should be "ondc.preprod.highwaydelite.com::ONDC:RET10".`  
- **Updated At Order**: `Message/issue/actions/3/updated_at should be less than or equal to message/issue/updated_at.`  

## issue(3)  
- **timestamp_validation**: `Issue updated_at should be equal to or greater than context timestamp.`  
- **Actor Org Name Mismatch**: `For actor type "INTERFACING_NP" in FLOW_2, org.name should be "ondc.preprod.highwaydelite.com::ONDC:RET10".`  
- **Updated At Order**: `Message/issue/actions/5/updated_at should be less than or equal to message/issue/updated_at.`  

## issue(7)  
- **Actor Org Name Mismatch**: `For actor type "INTERFACING_NP" in FLOW_2, org.name should be "ondc.preprod.highwaydelite.com::ONDC:RET10".`  
- **Updated At Order**: `Message/issue/actions/10/updated_at should be less than or equal to message/issue/updated_at.`

highwaydelite commented:
@amitsinha07 Updated the IGM logs. Please verify at your earliest convenience.

amitsinha07 commented:
@highwaydelite 
Flow 1 looks fine but in flow 2 why there are two issue apis with action OPEN (issue and issue_1)

highwaydelite commented:
@amitsinha07 Those were redundant files from previous commits that have now been deleted.

amitsinha07 commented:
@highwaydelite 

There are still some issues present in your logs. Kindly run the utility again.

Additionally, since the log validation utility has been updated, it’s important that you use the latest version to ensure accurate results.

**Could you please:**

1. Re-run the log validation utility using the updated version.
2. Raise a new pull request incorporating the changes.

This will allow me to review the logs using the latest standards and provide you with the most relevant feedback.

highwaydelite commented:
@92shreyansh 

IGM logs have been removed from this PR. We will create a new PR for IGM as instructed by @amitsinha07.

We request you to merge this PR as transaction flows for TRV10 and RSF have already been approved.
